Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm

```diff
@@ -239,6 +249,81 @@ void requestLinkDecorationFilteringData(LinkFilteringRulesCallback&& callback)
     }];
 }
 
+StorageAccessPromptQuirkController& StorageAccessPromptQuirkController::shared()
+{
+    static StorageAccessPromptQuirkController* sharedInstance = new StorageAccessPromptQuirkController;
+    return *sharedInstance;
+}
+
+Ref<StorageAccessPromptQuirkObserver> StorageAccessPromptQuirkController::observeUpdates(Function<void()>&& completionHandler)
+{
+    ASSERT(RunLoop::isMain());
+    if (!m_notificationListener) {
+        m_notificationListener = adoptNS([WKWebPrivacyNotificationListener new]);
+        [m_notificationListener listenForStorageAccessPromptQuirkChanges:^{
+            updateQuirks([this] {
+                for (auto& observer : m_observers)
```

Comment:
I think you want to use `WeakHashSet::forEach()` here to make a copy of the observers. It is a bit risky otherwise as an observer could unregister / register itself during the execution of the callback.